### PR TITLE
Update API

### PR DESCRIPTION
**Pull request by drasmuss**
Lots of changes to the LMU API to make it more user-friendly, and add support for new features (like multi-dimensional memories and dropout). 

Some of the most significant changes are:
- Removed a bunch of elements that we haven't really found to be that useful in our experimentation with LMUs, like individual trainable/initializer arguments for each connection within the LMU, or the LMUCellODE class (although we have plans to implement an improved method for optimizing the memory component in the future)
- Added the ability to selectively enable/disable connections within the LMU between the hidden and memory components. Note that these default to disabled, so the new default LMU will be a trimmed-down version of the old defaults.
- Added support for multi-dimensional memories.
- Added support for arbitrary hidden components (anything that implements the Keras RNNCell API can be used, e.g. `tf.keras.layers.LSTMCell`).

See the diff in CHANGES.rst for all the details.

### CODE REVIEW

**Review by gsmalik — COMMENTED** [review]
I had just one small comment about a potential dimension mismatch if `hidden_cell` is None and `input_to_hidden` is enabled.

File: lmu/layers.py [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+    ):
+        super().__init__(**kwargs)
+
+        self.memory_d = memory_d
+        self.order = order
+        self.theta = theta
+        self.hidden_cell = hidden_cell
+        self.hidden_to_memory = hidden_to_memory
+        self.memory_to_memory = memory_to_memory
+        self.input_to_hidden = input_to_hidden
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply lmu/layers.py:92]: I think there are valid reasons to have `input_to_hidden` with `hidden_cell=None` (e.g., you could use it as a way to do skip connections in an LMU stack), so I wouldn't want to disable it.

You're definitely right about the dimensions being wrong though, we need to add `inp_size` as you say. Could you do that, and also add a test for that case (since we're obviously missing that currently!).

[the remark this answers is not in the document]
Comment (gsmalik) [reply lmu/layers.py:92]: Yeah, I can try and do this. :+1:

[the remark this answers is not in the document]
Comment (gsmalik) [reply lmu/layers.py:92]: @drasmuss, I have added a fix in the [`build`](https://github.com/abr/lmu/blob/update-api/lmu/layers.py#L143-L145) part of LMUCell by updating the `hidden_output_size` and `output_size` if `hidden_cell` is None and `input_to_hidden` connection is enabled. 

For the tests, I have added a test that checks the dimensions of the output, `hidden_outout_size` and `output_size` when hidden cell is set (RNN/Dense) and not set with `inputt_to_hidden` connections enabled. 

Could you check quickly if this makes sense? I will then merge :+1:

[the remark this answers is not in the document]
Comment (drasmuss) [reply lmu/layers.py:92]: Pushed a minor change, where we set the output size to None before the layer is built (to make sure that anyone checking that value before building is aware that that value is not correct). And then just some small simplifications to the test. Looks good!

**Review by gsmalik — APPROVED** [review]
